File: Iverson.Clients/Java/client/src/main/java/io/iverson/client/core/EntityCoordinator.java
package io.iverson.client.core;

import io.grpc.StatusRuntimeException;
import io.grpc.stub.AbstractStub;
import io.iverson.client.annotations.IversonEntity;
import io.iverson.client.annotations.IversonKey;
import io.iverson.client.search.AggregateBuilder;
import io.iverson.client.search.ChunksBuilder;
import io.iverson.client.search.GroupByBuilder;
import io.iverson.client.search.PipelineBuilder;
import io.iverson.client.search.QueryBuilder;
import io.iverson.client.search.SimilarBuilder;
import iverson.ObjectMapping;
import iverson.ObjectMappingServiceGrpc;
import iverson.ObjectPersistence;
import iverson.ObjectRetrieval;
import iverson.ObjectSearch;
import iverson.ObjectSearchServiceGrpc;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Per-entity surface for all CRUD and search operations.
 *
 * <p>Uses the lightweight {@code ObjectPersistenceService} for writes and
 * {@code ObjectRetrievalService} for key-based reads. Search goes through
 * {@code ObjectSearchService}.</p>
 *
 * @param <T> the entity type; must be annotated with {@link IversonEntity}
 */
public final class EntityCoordinator<T> {

    private final IversonClient client;
    private final Class<T> entityType;
    private final String typeName;
    private final ObjectSearchServiceGrpc.ObjectSearchServiceBlockingStub searchStub;
    private final String boundActingUserToken;

    public EntityCoordinator(IversonClient client, Class<T> entityType) {
        if (entityType.getAnnotation(IversonEntity.class) == null) {
            throw new IllegalArgumentException(
                entityType.getSimpleName() + " is not annotated with @IversonEntity");
        }
        this.client               = client;
        this.entityType           = entityType;
        this.typeName             = entityType.getSimpleName();
        this.searchStub           = client.searchStub;
        this.boundActingUserToken = null;
        // Validate that a key field exists
        findKeyField(entityType);
    }

    /** Package-private constructor for testing with a mock search stub. */
    EntityCoordinator(ObjectSearchServiceGrpc.ObjectSearchServiceBlockingStub searchStub, Class<T> entityType) {
        if (entityType.getAnnotation(IversonEntity.class) == null) {
            throw new IllegalArgumentException(
                entityType.getSimpleName() + " is not annotated with @IversonEntity");
        }
        this.client               = null;
        this.entityType           = entityType;
        this.typeName             = entityType.getSimpleName();
        this.searchStub           = searchStub;
        this.boundActingUserToken = null;
        // Validate that a key field exists
        findKeyField(entityType);
    }

    /** Copy constructor backing {@link #withActingUser(String)}. {@code client} may be null. */
    private EntityCoordinator(IversonClient client,
                               ObjectSearchServiceGrpc.ObjectSearchServiceBlockingStub searchStub,
                               Class<T> entityType,
                               String boundActingUserToken) {
        this.client               = client;
        this.entityType           = entityType;
        this.typeName             = entityType.getSimpleName();
        this.searchStub           = searchStub;
        this.boundActingUserToken = boundActingUserToken;
    }

    /**
     * Returns a copy of this coordinator bound to the given acting-user identity. The bound
     * token applies to every subsequent call on the returned coordinator that supplies no more
     * specific (per-call explicit) token, taking precedence over the client's ambient identity.
     */
    public EntityCoordinator<T> withActingUser(String actingUserToken) {
        return new EntityCoordinator<>(client, searchStub, entityType, actingUserToken);
    }

    // ── Object Persistence (lightweight writes) ────────────────────────────────

    /**
     * Persists a new entity and returns the server-assigned key.
     */
    public String persist(T entity) throws StatusRuntimeException {
        ObjectPersistence.PersistRequest request = ObjectPersistence.PersistRequest.newBuilder()
            .setTypeName(typeName)
            .setPayload(StructConverter.toStruct(entity))
            .build();
        ObjectPersistence.PersistResponse response = withIdentity(client.persistenceStub, null).post(request);
        if (!response.getSuccess()) {
            throw new StatusRuntimeException(
                io.grpc.Status.INTERNAL.withDescription(response.getError()));
        }
        return response.getKey();
    }

    /**
     * Updates an existing entity (matched by the key field).
     */
    public void update(T entity) throws StatusRuntimeException {
        ObjectPersistence.PersistRequest request = ObjectPersistence.PersistRequest.newBuilder()
            .setTypeName(typeName)
            .setPayload(StructConverter.toStruct(entity))
            .build();
        ObjectPersistence.PersistResponse response = withIdentity(client.persistenceStub, null).update(request);
        if (!response.getSuccess()) {
            throw new StatusRuntimeException(
                io.grpc.Status.INTERNAL.withDescription(response.getError()));
        }
    }

    // ── Object Retrieval ───────────────────────────────────────────────────────

    /**
     * Fetches a single entity by its key. Returns {@code null} if not found.
     */
    public T get(String id) throws StatusRuntimeException {
        ObjectRetrieval.RetrievalRequest request = ObjectRetrieval.RetrievalRequest.newBuilder()
            .setTypeName(typeName)
            .setKey(id)
            .build();
        ObjectRetrieval.RetrievalResponse response = withIdentity(client.retrievalStub, null).get(request);
        if (!response.getFound()) return null;
        return StructConverter.fromStruct(response.getData(), entityType);
    }

    /**
     * Fetches multiple entities by their keys. Missing entities are silently omitted.
     */
    public List<T> getMany(List<String> ids) throws StatusRuntimeException {
        ObjectRetrieval.RetrievalManyRequest request = ObjectRetrieval.RetrievalManyRequest.newBuilder()
            .setTypeName(typeName)
            .addAllKeys(ids)
            .build();
        Iterator<ObjectRetrieval.RetrievalResponse> stream = withIdentity(client.retrievalStub, null).getMany(request);
        List<T> results = new ArrayList<>();
        while (stream.hasNext()) {
            ObjectRetrieval.RetrievalResponse response = stream.next();
            if (!response.getFound()) continue;
            T entity = StructConverter.fromStruct(response.getData(), entityType);
            if (entity != null) results.add(entity);
        }
        return results;
    }

    /**
     * Deletes the entity with the given key.
     */
    public void delete(String id) throws StatusRuntimeException {
        ObjectMapping.MappingDeleteRequest request =
            ObjectMapping.MappingDeleteRequest.newBuilder()
                .setTypeName(typeName)
                .setKey(id)
                .build();
        ObjectMapping.MappingDeleteResponse response =
            withIdentity(client.mappingStub, null).delete(request);
        if (!response.getSuccess()) {
            throw new StatusRuntimeException(
                io.grpc.Status.INTERNAL.withDescription(response.getError()));
        }
    }

    // ── Object Mapping (full CRUD with relation resolution) ────────────────────

    /**
     * Fetches a single entity by key with server-side relation resolution to {@code depth}.
     * Returns {@code null} if not found.
     */
    public T getMapped(String id, int depth, String actingUserToken) throws StatusRuntimeException {
        ObjectMapping.MappingGetRequest request = ObjectMapping.MappingGetRequest.newBuilder()
            .setTypeName(typeName)
            .setKey(id)
            .setDepth(depth)
            .build();
        ObjectMapping.MappingResponse response = mappingStubFor(actingUserToken).get(request);
        if (!response.getSuccess()) return null;
        return StructConverter.fromStruct(response.getData(), entityType);
    }

    /**
     * Creates an entity through the mapping path, which resolves its relations server-side.
     * Returns the entity hydrated from the response, carrying the server-assigned key — the
     * caller never assigns one.
     */
    public T postMapped(T entity, String actingUserToken) throws StatusRuntimeException {
        ObjectMapping.MappingWriteRequest request = ObjectMapping.MappingWriteRequest.newBuilder()
            .setTypeName(typeName)
            .setPayload(StructConverter.toStruct(entity))
            .build();
        ObjectMapping.MappingResponse response = mappingStubFor(actingUserToken).post(request);
        if (!response.getSuccess()) {
            throw new StatusRuntimeException(
                io.grpc.Status.INTERNAL.withDescription(response.getError()));
        }
        return StructConverter.fromStruct(response.getData(), entityType);
    }

    /** Updates an existing entity through the mapping path. */
    public T updateMapped(T entity, String actingUserToken) throws StatusRuntimeException {
        ObjectMapping.MappingWriteRequest request = ObjectMapping.MappingWriteRequest.newBuilder()
            .setTypeName(typeName)
            .setPayload(StructConverter.toStruct(entity))
            .build();
        ObjectMapping.MappingResponse response = mappingStubFor(actingUserToken).update(request);
        if (!response.getSuccess()) {
            throw new StatusRuntimeException(
                io.grpc.Status.INTERNAL.withDescription(response.getError()));
        }
        return StructConverter.fromStruct(response.getData(), entityType);
    }

    // ── Object Search ──────────────────────────────────────────────────────────

    /**
     * Executes a search query and returns all matching results as a list.
     */
    public List<SearchResult<T>> search(QueryBuilder<T> queryBuilder) throws StatusRuntimeException {
        ObjectSearch.SearchRequest request = queryBuilder.build();
        Iterator<ObjectSearch.SearchResponse> stream = stubFor(null).search(request);
        List<SearchResult<T>> results = new ArrayList<>();
        while (stream.hasNext()) {
            ObjectSearch.SearchResponse response = stream.next();
            T entity = StructConverter.fromStruct(response.getData(), entityType);
            if (entity != null) results.add(new SearchResult<>(entity, response.getScore()));
        }
        return results;
    }

    /**
     * Executes a compound GROUP BY aggregation and returns one row per output group. Column
     * set depends on the query's keys/metrics, so rows come back as string-keyed maps rather
     * than typed entities.
     */
    public List<Map<String, Object>> groupBy(GroupByBuilder builder) throws StatusRuntimeException {
        return groupBy(builder, null);
    }

    /** Same as {@link #groupBy(GroupByBuilder)}, propagating an acting-user token if given. */
    public List<Map<String, Object>> groupBy(GroupByBuilder builder, String actingUserToken)
            throws StatusRuntimeException {
        ObjectSearch.GroupByRequest request = builder.build();
        Iterator<ObjectSearch.SearchResponse> stream = stubFor(actingUserToken).groupBy(request);
        List<Map<String, Object>> results = new ArrayList<>();
        while (stream.hasNext()) {
            results.add(StructConverter.fromStructAsMap(stream.next().getData()));
        }
        return results;
    }

    /**
     * Executes an aggregation request and returns the full {@link ObjectSearch.AggregateResponse}
     * (one {@code AggregationResult} per requested {@code AggregationSpec}).
     */
    public ObjectSearch.AggregateResponse aggregate(AggregateBuilder builder) throws StatusRuntimeException {
        return aggregate(builder, null);
    }

    /** Same as {@link #aggregate(AggregateBuilder)}, propagating an acting-user token if given. */
    public ObjectSearch.AggregateResponse aggregate(AggregateBuilder builder, String actingUserToken)
            throws StatusRuntimeException {
        ObjectSearch.AggregateRequest request = builder.build();
        return stubFor(actingUserToken).aggregate(request);
    }

    /**
     * Executes a pipeline (CTE chain) and returns one row per output row. Column set depends
     * on the pipeline's final step, so rows come back as string-keyed maps, same as
     * {@link #groupBy(GroupByBuilder)}.
     */
    public List<Map<String, Object>> pipeline(PipelineBuilder builder) throws StatusRuntimeException {
        return pipeline(builder, null);
    }

    /** Same as {@link #pipeline(PipelineBuilder)}, propagating an acting-user token if given. */
    public List<Map<String, Object>> pipeline(PipelineBuilder builder, String actingUserToken)
            throws StatusRuntimeException {
        ObjectSearch.PipelineRequest request = builder.build();
        Iterator<ObjectSearch.SearchResponse> stream = stubFor(actingUserToken).pipeline(request);
        List<Map<String, Object>> results = new ArrayList<>();
        while (stream.hasNext()) {
            results.add(StructConverter.fromStructAsMap(stream.next().getData()));
        }
        return results;
    }

    /** Executes a Qdrant vector similarity search and returns matching entities with scores. */
    public List<SearchResult<T>> searchSimilar(SimilarBuilder builder) throws StatusRuntimeException {
        return searchSimilar(builder, null);
    }

    /** Same as {@link #searchSimilar(SimilarBuilder)}, propagating an acting-user token if given. */
    public List<SearchResult<T>> searchSimilar(SimilarBuilder builder, String actingUserToken)
            throws StatusRuntimeException {
        ObjectSearch.SearchSimilarRequest request = builder.build();
        Iterator<ObjectSearch.SearchResponse> stream = stubFor(actingUserToken).searchSimilar(request);
        List<SearchResult<T>> results = new ArrayList<>();
        while (stream.hasNext()) {
            ObjectSearch.SearchResponse response = stream.next();
            T entity = StructConverter.fromStruct(response.getData(), entityType);
            if (entity != null) results.add(new SearchResult<>(entity, response.getScore()));
        }
        return results;
    }

    /**
     * Executes a Qdrant chunk/RAG search and returns matching passage chunks with their
     * parent entity key and relevance score.
     */
    public List<ChunkSearchResult> searchChunks(ChunksBuilder builder) throws StatusRuntimeException {
        return searchChunks(builder, null);
    }

    /** Same as {@link #searchChunks(ChunksBuilder)}, propagating an acting-user token if given. */
    public List<ChunkSearchResult> searchChunks(ChunksBuilder builder, String actingUserToken)
            throws StatusRuntimeException {
        ObjectSearch.SearchChunksRequest request = builder.build();
        Iterator<ObjectSearch.ChunkSearchResponse> stream = stubFor(actingUserToken).searchChunks(request);
        List<ChunkSearchResult> results = new ArrayList<>();
        while (stream.hasNext()) {
            ObjectSearch.ChunkSearchResponse response = stream.next();
            results.add(new ChunkSearchResult(response.getParentKey(), response.getChunkText(), response.getScore()));
        }
        return results;
    }

    /**
     * Attaches the resolved acting-user identity to {@code stub} as a call option (consumed by
     * {@link OAuth2ClientCredentials}). Resolution order: the caller's explicit token, then this
     * coordinator's bound identity, then the client's ambient one; none attaches nothing.
     */
    private <S extends AbstractStub<S>> S withIdentity(S stub, String explicitToken) {
        String token = explicitToken != null ? explicitToken
            : boundActingUserToken != null ? boundActingUserToken
            : (client != null ? client.actingUserToken : null);
        return token != null ? stub.withOption(OAuth2ClientCredentials.ACTING_USER_TOKEN, token) : stub;
    }

    /**
     * Returns the search stub to invoke, attaching the resolved acting-user identity as a call
     * option (consumed by {@link OAuth2ClientCredentials}).
     */
    private ObjectSearchServiceGrpc.ObjectSearchServiceBlockingStub stubFor(String actingUserToken) {
        return withIdentity(searchStub, actingUserToken);
    }

    /**
     * Returns the mapping stub to invoke, attaching the resolved acting-user identity as a call
     * option (consumed by {@link OAuth2ClientCredentials}). The constructor's
     * {@link io.grpc.CallCredentials} are the service's own client-credentials token and do
     * <em>not</em> identify an acting user; the server denies any write without one.
     */
    private ObjectMappingServiceGrpc.ObjectMappingServiceBlockingStub mappingStubFor(String actingUserToken) {
        return withIdentity(client.mappingStub, actingUserToken);
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    private static Field findKeyField(Class<?> cls) {
        for (Field field : getAllFields(cls)) {
            if (field.getAnnotation(IversonKey.class) != null) {
                field.setAccessible(true);
                return field;
            }
        }
        throw new IllegalArgumentException(
            cls.getSimpleName() + " has no field annotated with @IversonKey");
    }

    private static List<Field> getAllFields(Class<?> cls) {
        List<Field> fields = new ArrayList<>();
        while (cls != null && cls != Object.class) {
            for (Field f : cls.getDeclaredFields()) {
                if (!f.isSynthetic()) fields.add(f);
            }
            cls = cls.getSuperclass();
        }
        return fields;
    }

    /** Wraps a search result entity with its relevance score. */
    public record SearchResult<T>(T entity, float score) {}

    /** Wraps a chunk/RAG search hit with its parent entity key, passage text, and relevance score. */
    public record ChunkSearchResult(String parentKey, String chunkText, float score) {}
}

File: Iverson.Clients/DotNet/Iverson.Client.Core/EntityCoordinator.cs
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Grpc.Core;
using Iverson.Client.Contracts;
using Iverson.Client.Search;
using Microsoft.Extensions.Logging;

namespace Iverson.Client.Core;

/// <summary>
/// The single surface developers use to interact with Iverson entities.
/// Routes each operation to the correct gRPC service; completely ignorant of
/// what backing stores the server uses.
/// </summary>
public sealed class EntityCoordinator<T>(
    EntityRegistry registry,
    GraphAssembler assembler,
    ObjectMappingService.ObjectMappingServiceClient mapping,
    ObjectPersistenceService.ObjectPersistenceServiceClient persistence,
    ObjectRetrievalService.ObjectRetrievalServiceClient retrieval,
    ObjectSearchService.ObjectSearchServiceClient search,
    ILogger<EntityCoordinator<T>> logger,
    ActingUserIdentity? identity = null)
    where T : class
{
    private readonly EntityDescriptor _descriptor = registry.Get<T>();

    private IReadOnlyCollection<string> NavPropertyNames =>
        _descriptor.Relations.Select(r => r.Property.Name).ToList();

    // Deliberately not `readonly`: WithActingUser assigns it through an object initializer on a
    // freshly constructed instance, which the compiler rejects on a readonly field. Nothing else
    // ever writes it, so the instance is still effectively immutable after construction.
    private Func<Task<string>>? _boundActingUser;

    /// <summary>
    /// Returns a coordinator bound to <paramref name="tokenProvider"/>, leaving this one
    /// untouched. The bound identity outranks the client's ambient one; explicit headers
    /// carrying an acting-user entry outrank both.
    /// </summary>
    public EntityCoordinator<T> WithActingUser(Func<Task<string>> tokenProvider) =>
        new(registry, assembler, mapping, persistence, retrieval, search, logger, identity)
        {
            _boundActingUser = tokenProvider,
        };

    private async Task<Metadata> ResolveHeadersAsync(Metadata? headers)
    {
        headers ??= new Metadata();
        if (headers.Get(ActingUserMetadata.MetadataKey) is not null)
            return headers;

        var provider = _boundActingUser ?? identity?.TokenProvider;
        if (provider is null)
            return headers;

        // Never write into the caller's own Metadata instance: build a fresh copy so a bag the
        // caller reuses across calls (e.g. on different coordinators bound to different
        // identities) isn't mutated out from under them.
        var resolved = new Metadata();
        foreach (var entry in headers)
            resolved.Add(entry.IsBinary
                ? new Metadata.Entry(entry.Key, entry.ValueBytes)
                : new Metadata.Entry(entry.Key, entry.Value));
        resolved.WithActingUser(await provider());

        return resolved;
    }

    // ── Object Mapping ─────────────────────────────────────────────────────────
    // Full CRUD with server-side relationship resolution.

    public async Task<T?> GetMappedAsync(
        string key, int depth = 1, Metadata? headers = null, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectMapping.Get {Entity}:{Key} depth={Depth}", _descriptor.EntityName, key, depth);
        var response = await mapping.GetAsync(
            new MappingGetRequest
            {
                TypeName = _descriptor.EntityName,
                Key      = key,
                Depth    = depth,
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(headers), cancellationToken: ct);

        if (!response.Success) { LogError(response.Error); return null; }
        return StructConverter.FromStruct<T>(response.Data);
    }

    public async Task<T?> PostMappedAsync(T entity, Metadata? headers = null, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectMapping.Post {Entity}", _descriptor.EntityName);
        var response = await mapping.PostAsync(
            new MappingWriteRequest
            {
                TypeName = _descriptor.EntityName,
                Payload  = StructConverter.ToStruct(entity, NavPropertyNames),
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(headers), cancellationToken: ct);

        if (!response.Success) { LogError(response.Error); return null; }
        return StructConverter.FromStruct<T>(response.Data);
    }

    public async Task<T?> UpdateMappedAsync(T entity, Metadata? headers = null, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectMapping.Update {Entity}", _descriptor.EntityName);
        var response = await mapping.UpdateAsync(
            new MappingWriteRequest
            {
                TypeName = _descriptor.EntityName,
                Payload  = StructConverter.ToStruct(entity, NavPropertyNames),
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(headers), cancellationToken: ct);

        if (!response.Success) { LogError(response.Error); return null; }
        return StructConverter.FromStruct<T>(response.Data);
    }

    public async Task<bool> DeleteAsync(string key, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectMapping.Delete {Entity}:{Key}", _descriptor.EntityName, key);
        var response = await mapping.DeleteAsync(
            new MappingDeleteRequest
            {
                TypeName = _descriptor.EntityName,
                Key      = key,
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(null), cancellationToken: ct);

        if (!response.Success) LogError(response.Error);
        return response.Success;
    }

    // ── Object Persistence ─────────────────────────────────────────────────────
    // Lightweight writes — no relation traversal on the server.

    public async Task<string?> PersistAsync(T entity, Metadata? headers = null, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectPersistence.Post {Entity}", _descriptor.EntityName);
        var response = await persistence.PostAsync(
            new PersistRequest
            {
                TypeName = _descriptor.EntityName,
                Payload  = StructConverter.ToStruct(entity, NavPropertyNames),
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(headers), cancellationToken: ct);

        if (!response.Success) { LogError(response.Error); return null; }
        return response.Key;
    }

    public async Task<string?> UpdateAsync(T entity, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectPersistence.Update {Entity}", _descriptor.EntityName);
        var response = await persistence.UpdateAsync(
            new PersistRequest
            {
                TypeName = _descriptor.EntityName,
                Payload  = StructConverter.ToStruct(entity, NavPropertyNames),
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(null), cancellationToken: ct);

        if (!response.Success) { LogError(response.Error); return null; }
        return response.Key;
    }

    // ── Object Retrieval ───────────────────────────────────────────────────────
    // Key-based fetch; client assembles the graph locally from relation metadata.

    public async Task<T?> GetAsync(string key, bool assembleGraph = true, CancellationToken ct = default)
    {
        logger.LogDebug("ObjectRetrieval.Get {Entity}:{Key}", _descriptor.EntityName, key);
        var response = await retrieval.GetAsync(
            new RetrievalRequest
            {
                TypeName = _descriptor.EntityName,
                Key      = key,
                TraceId  = CurrentTraceId()
            },
            await ResolveHeadersAsync(null), cancellationToken: ct);

        if (!response.Found) return null;

        var entity = StructConverter.FromStruct<T>(response.Data);
        if (entity is not null && assembleGraph)
            await assembler.AssembleAsync(entity, ct);

        return entity;
    }

    public async IAsyncEnumerable<T> GetManyAsync(
        IEnumerable<string> keys,
        bool assembleGraph = true,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request = new RetrievalManyRequest { TypeName = _descriptor.EntityName, TraceId = CurrentTraceId() };
        request.Keys.AddRange(keys);

        logger.LogDebug("ObjectRetrieval.GetMany {Entity} ({Count} keys)", _descriptor.EntityName, request.Keys.Count);

        // Buffer all results so we can batch-assemble relations in one pass
        var buffer = new List<T>();
        var stream  = retrieval.GetMany(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
        {
            if (!response.Found) continue;
            var entity = StructConverter.FromStruct<T>(response.Data);
            if (entity is null) continue;
            buffer.Add(entity);
        }

        if (assembleGraph && buffer.Count > 0)
            await assembler.AssembleManyAsync(buffer, ct);

        foreach (var entity in buffer)
            yield return entity;
    }

    // ── Object Search ──────────────────────────────────────────────────────────
    // DSL-driven; returns streamed results with relevance scores.

    public async IAsyncEnumerable<SearchResult<T>> SearchAsync(
        QueryBuilder<T> query,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request     = query.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.Search {Entity} ({Clauses} clauses)",
            _descriptor.EntityName, request.Query?.Clauses.Count ?? 0);

        var stream = search.Search(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
        {
            var entity = StructConverter.FromStruct<T>(response.Data);
            if (entity is not null)
                yield return new SearchResult<T>(entity, response.Score);
        }
    }

    public async IAsyncEnumerable<SearchResult<T>> SearchSimilarAsync(
        QuerySimilarBuilder<T> query,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request     = query.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.SearchSimilar {Entity} property={Property}",
            _descriptor.EntityName, request.Property);

        var stream = search.SearchSimilar(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
        {
            var entity = StructConverter.FromStruct<T>(response.Data);
            if (entity is not null) yield return new SearchResult<T>(entity, response.Score);
        }
    }

    public async IAsyncEnumerable<ChunkSearchResponse> SearchChunksAsync(
        QueryChunksBuilder<T> query,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request     = query.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.SearchChunks {Entity} property={Property}",
            _descriptor.EntityName, request.Property);

        var stream = search.SearchChunks(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
            yield return response;
    }

    /// <summary>
    /// Executes a pipeline (CTE chain) and streams untyped rows. Column set depends on the
    /// pipeline's final step, so rows come back as string-keyed dictionaries.
    /// </summary>
    public async IAsyncEnumerable<IReadOnlyDictionary<string, object?>> PipelineAsync(
        PipelineBuilder pipeline,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request     = pipeline.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.Pipeline {Entity} ({Steps} steps)",
            _descriptor.EntityName, request.Steps.Count);

        var stream = search.Pipeline(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
            yield return StructConverter.ToDictionary(response.Data);
    }

    /// <summary>
    /// Executes a pipeline and projects each row onto <typeparamref name="TResult"/>
    /// (any class whose property names match the pipeline's output aliases).
    /// </summary>
    public async IAsyncEnumerable<TResult> PipelineAsync<TResult>(
        PipelineBuilder pipeline,
        [EnumeratorCancellation] CancellationToken ct = default)
        where TResult : class
    {
        var request     = pipeline.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.Pipeline {Entity} ({Steps} steps, typed)",
            _descriptor.EntityName, request.Steps.Count);

        var stream = search.Pipeline(request, await ResolveHeadersAsync(null), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
        {
            var row = StructConverter.FromStruct<TResult>(response.Data);
            if (row is not null) yield return row;
        }
    }

    /// <summary>
    /// Executes a GROUP BY aggregation and streams untyped rows (one row per output group).
    /// Column set depends on the query's keys/metrics, so rows come back as string-keyed
    /// dictionaries, same as <see cref="PipelineAsync(PipelineBuilder,CancellationToken)"/>.
    /// </summary>
    public async IAsyncEnumerable<IReadOnlyDictionary<string, object?>> GroupByAsync(
        GroupByBuilder query,
        Metadata? headers = null,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var request     = query.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.GroupBy {Entity} ({Keys} keys)",
            _descriptor.EntityName, request.Keys.Count);

        var stream = search.GroupBy(request, await ResolveHeadersAsync(headers), cancellationToken: ct);
        await foreach (var response in stream.ResponseStream.ReadAllAsync(ct))
            yield return StructConverter.ToDictionary(response.Data);
    }

    /// <summary>
    /// Executes an aggregation request and returns the full <see cref="AggregateResponse"/>
    /// (one <see cref="AggregationResult"/> per requested <see cref="AggregationSpec"/>).
    /// </summary>
    public async Task<AggregateResponse> AggregateAsync(
        AggregateBuilder query,
        Metadata? headers = null,
        CancellationToken ct = default)
    {
        var request     = query.Build();
        request.TraceId = CurrentTraceId();

        logger.LogDebug("ObjectSearch.Aggregate {Entity} ({Aggregations} aggregations)",
            _descriptor.EntityName, request.Aggregations.Count);

        return await search.AggregateAsync(request, await ResolveHeadersAsync(headers), cancellationToken: ct);
    }

    private static string CurrentTraceId() =>
        Activity.Current?.TraceId.ToString() ?? string.Empty;

    private void LogError(string error) =>
        logger.LogError("Iverson gRPC error for {Entity}: {Error}", _descriptor.EntityName, error);
}

public sealed record SearchResult<T>(T Entity, float Score);
